tests/test_failures.py

```diff
@@ -116,3 +118,35 @@ def test_transaction(var: str) -> str:
     dbos._sys_db.wait_for_buffer_flush()
     wfs = dbos._sys_db.get_workflows(gwi)
     assert len(wfs.workflow_uuids) == 2
+
+
+def test_dead_letter_queue(dbos: DBOS) -> None:
+    event = threading.Event()
+    max_recovery_attempts = 20
+    recovery_count = 0
+
+    @DBOS.workflow(max_recovery_attempts=max_recovery_attempts)
+    def dead_letter_workflow() -> None:
+        nonlocal recovery_count
+        recovery_count += 1
+        event.wait()
+
+    handle = DBOS.start_workflow(dead_letter_workflow)
+
+    for i in range(max_recovery_attempts):
+        DBOS.recover_pending_workflows()
+        assert recovery_count == i + 2
+
+    with pytest.raises(Exception) as exc_info:
+        DBOS.recover_pending_workflows()
+    assert exc_info.errisinstance(DBOSDeadLetterQueueError)
+    assert handle.get_status().status == WorkflowStatusString.RETRIES_EXCEEDED.value
+
+    with SetWorkflowID(handle.get_workflow_id()):
+        DBOS.start_workflow(dead_letter_workflow)
```

Comment:
It's not obvious that start_workflow again will resume the workflow. We need to document this behavior...